### PR TITLE
[inductor] move all cpu scalars using pinned memory for graph partition (#155360)

### PR DESCRIPTION
**Pull request by v0i0**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158983
* #158758

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

### CODE REVIEW

File: torch/_inductor/fx_passes/post_grad.py [hunk 1]
```diff
@@ -1720,17 +1720,45 @@ def __call__(self, graph: fx.Graph) -> None:
...
         target_device = next(iter(target_devices))
-        for node in movable_constructors:
-            if node in cpu_placeholders:
-                with graph.inserting_after(node):
-                    gpu_node = graph.call_function(
-                        torch.ops.prims.device_put.default, (node, target_device)
+        movable_cpu_placeholders = movable_constructors & cpu_placeholders
+        if movable_cpu_placeholders:
+            for node in movable_cpu_placeholders:
+                pass
```
Comment (eellison) [inline torch/_inductor/fx_passes/post_grad.py]: nit: `node = next(iter(reversed(movable_cpu_placeholders)))`

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -1720,17 +1720,45 @@ def __call__(self, graph: fx.Graph) -> None:
...
+            last_node = node
+            unsqueezed_nodes = []
+            for elem in movable_cpu_placeholders:
+                with graph.inserting_after(last_node):
+                    unsqueezed_nodes.append(
+                        graph.call_function(torch.ops.aten.unsqueeze.default, (elem, 0))
```
Comment (eellison) [inline torch/_inductor/fx_passes/post_grad.py:1771]: I guess cat doesn't accept 0-dim tensors..

File: torch/_inductor/fx_passes/post_grad.py [hunk 3]
```diff
@@ -1720,17 +1720,45 @@ def __call__(self, graph: fx.Graph) -> None:
...
+                    torch.ops.prims.device_put.default,
+                    (cpu_concat, target_device, True),
                 )
+                last_node = gpu_concat
+            with graph.inserting_after(last_node):
+                gpu_split = graph.call_function(
+                    torch.ops.aten.unbind.int, (gpu_concat,)
+                )
+                last_node = gpu_split
+            for idx, node in enumerate(movable_cpu_placeholders):
```
Comment (eellison) [inline torch/_inductor/fx_passes/post_grad.py:1790]: So, just as an fyi, typically passes maintain the meta of each node which contains the properties of each tensor. the reason we dont need to do that in this pass is because we directly run an update pass after.

https://github.com/pytorch/pytorch/blob/afd6eb0d495a9e20bbda26a92312521cd408c13a/torch/_inductor/fx_passes/post_grad.py#L187-L191

**Review by eellison — APPROVED** [review]
looks good !!

File: torch/_inductor/fx_passes/post_grad.py (hunk 2 above)
Comment (v0i0) [reply torch/_inductor/fx_passes/post_grad.py:1771]: yes we'd want torch.stack, but that doesn't have a prim
